### PR TITLE
connecting mocks to spectra

### LINKED ISSUE #88: Encode mock provenance into TARGETID.

### LINKED ISSUE #89: Explore the possibility to split the monolithic Targets files into sweeps/bricks [mocks]

### PR DESCRIPTION
**Pull request by moustakas**
This PR contains the long-overdue refactor of the code which connects the mocks to spectra (see the long thread in #136 for many relevant details).  It also closes #89, #133, #149, and (partially) #130.  I'm submitting it now even though it's not ready to merge (for example it's leaning on branches of `desisim` and `desiutil`!) so that a review by @forero,  @sbailey, and/or @geordie666 can begin.  

I have not yet tested this at NERSC, but on my laptop I was able to generate truth and target files (including spectra) for BGS, ELG, LRG, QSO, LYA, MWS_WD, MWD_NEARBY, MWS_MAIN (which includes F-type standards), and SKY targets (using the same input yaml file) over 140 bricks (~9 deg2) in ~28 min using 4 cores, which is not bad (although this will slow down because of https://github.com/desihub/desisim/issues/209).  Two of the important next steps are to profile the code and to also make sure everything scales over the full footprint.

Besides producing a relatively realistic set of spectra for all the mock targets, some of the other **major developments** include:
* Results are (should be!) reproducible based on an input random number seed, which is written out into the header of the `truth.fits` metadata table.
* Spectra are assigned in a physically motivated way using a fast KD Tree method (for most object types; shout-out to @dkirkby) and fluxes/colors/shapes are assigned using @belaa's Gaussian Mixture Models, where needed.
* Assigning spectra and writing out files is parallelized using the `multiprocessing` module.
* The `targets.fits` file is minimalistic, i.e., it is very similar to a targeting catalog, and it includes "noisy" *grzW1W2* photometry (to complement the noiseless photometry in the corresponding `truth.fits` files).
* Catalogs + noiseless spectra are written out to arbitrarily sized bricks (currently 2x2 deg2) although the work is done on the 0.25x0.25 deg2 bricks used by the imaging surveys (to accommodate the target density fluctuations).
* Lots of new and reorganized code (e.g., new Classes) and informational log messages. Also the target selection is done with `desitarget.cuts` to the greatest extent possible.
* The input mock yaml file has been simplified significantly, which should help accommodate future mocks.

Some of the **highest-priority ToDo** items include (in no particular order):
* Add shapes and fluxes to the output catalogs (this is easy, just hasn't risen to the top).
* Track down why no standard stars are being selected and written out (likely a bug).
* Include velocity dispersion broadening in the galaxy spectra (temporarily suppressed for speed -- see https://github.com/desihub/desisim/issues/209).
* Add contaminants.  With the new data model we should be able to include several different types of contaminants for several different types of objects (e.g., contaminating `STARs` and `ELGs` for `QSOs`, contaminating `STARs` for `ELGs`, etc.) relatively easily.
* Validate the space-densities of certain spectral classes in the mocks.  For example, the `SKY`, `QSO`, and `ELG` Gaussian Random Field mocks appear to be too sparse after targeting selection has been applied.  [@fjaviersanchez, @dkirkby I will try to get you more information on this.]
* Add some basic unit tests.

There are also a number of **known simplifications** that I'll document later.  I would also like to try to close #88 (encode mock provenance into `TARGETID`).  @forero, @sbailey any thoughts on how you would like this to be done?

### CODE REVIEW

**Comment by forero** [comment]
This is great! Do you have a list of commands to execute (+examples of input files) to produce some outputs?

**Comment by moustakas** [comment]
I took the liberty of creating the directory
```
/global/project/projectdirs/desi/datachallenge/fullsurvey2017
```
in which I put a file called `mock_moustakas.yaml`.  The code won't run until the dependencies have been sorted out (for example, the basis_templates need to be updated to `v2.3`, upon which a pending `desisim` PR is waiting for), but once everything is resolved you can run with, e.g.,
```
time select_mock_targets -c mock_moustakas.yaml -s 678 --nproc 4 --output_dir test
```

**Comment by forero** [comment]
@moustakas Running on edison I get this error

```
INFO:build.py:529:targets_truth: Assigned 971 QSO objects to 104 unique 0.25x0.25 deg2 bricks.
INFO:build.py:536:_update_spectra_status: 10 bricks; 2.4 sec / brick
INFO:build.py:536:_update_spectra_status: 20 bricks; 2.3 sec / brick
INFO:build.py:536:_update_spectra_status: 30 bricks; 2.2 sec / brick
INFO:build.py:536:_update_spectra_status: 40 bricks; 2.2 sec / brick
INFO:build.py:536:_update_spectra_status: 50 bricks; 2.0 sec / brick
INFO:build.py:536:_update_spectra_status: 60 bricks; 1.9 sec / brick
INFO:build.py:536:_update_spectra_status: 70 bricks; 2.0 sec / brick
INFO:build.py:536:_update_spectra_status: 80 bricks; 1.9 sec / brick
INFO:build.py:536:_update_spectra_status: 90 bricks; 1.8 sec / brick
INFO:build.py:536:_update_spectra_status: 100 bricks; 1.9 sec / brick
INFO:build.py:577:targets_truth: Downsampling to desired target density 170 targets/deg2.
Traceback (most recent call last):
  File "/global/homes/f/forero/desitarget/bin/select_mock_targets", line 53, in <module>
    outbricksize=args.outbricksize)
  File "/global/homes/f/forero/desitarget/py/desitarget/mock/build.py", line 579, in targets_truth
    sourcename=source_name)
  File "/global/homes/f/forero/desitarget/py/desitarget/mock/selection.py", line 256, in density_select
    keep.append(self.rand.choice(onbrick, int(n_in_brick * frac_keep), replace=False))
TypeError: only length-1 arrays can be converted to Python scalars
```

any idea of why?

File: py/desitarget/mock/io.py [hunk 1]
```diff
@@ -33,211 +38,106 @@
 ENCODE_FILE_MASK   = 2**ENCODE_FILE_END - 2**ENCODE_ROW_END
 ENCODE_FILE_MAX    = ENCODE_FILE_MASK >> ENCODE_ROW_END
 
-
+C_LIGHT = 299792.458
```
Comment (sbailey) [inline py/desitarget/mock/io.py]: Minor: preferable to use `astropy.constants.c.to('km/s').value` or `scipy.constants.c/1000`.

**Comment by moustakas** [comment]
(I'm working on repairing the failing tests and had already added speclite as a dependency, BTW...)

**Comment by moustakas** [comment]
@forero OK, I think I fixed the issue you identified and I've been able to successfully run the code at NERSC:
```
cd /scratch1/scratchdirs/ioannis/datachallenge2017
time select_mock_targets -c mock_input.yaml -s 678 --nproc 24 --output_dir test
```
The software stack that gets loaded by
```
source /project/projectdirs/desi/software/desi_environment.sh
```
is out of date (even when I load the master versions of the repositories, which I don't understand), so you'll need to clone master or the latest tag of desiutil (1.9.3), desisim (0.17.2), and desispec (master -- we need a tag), as well as the `spectra` branch of `desitarget`.  

```
[snip]

INFO:build.py:622:targets_truth: Writing test/sky.fits
INFO:build.py:625:targets_truth: Removing 8618 SKY targets from targets, truth, and trueflux.

INFO:build.py:644:targets_truth: Writing test/standards-dark.fits
INFO:build.py:644:targets_truth: Writing test/standards-bright.fits
INFO:build.py:653:targets_truth: Writing out 26595 targets to 2 2.0x2.0 deg2 bricks.
INFO:build.py:440:write_onebrick: Writing test/209/truth-2090m020.fits.
INFO:build.py:440:write_onebrick: Writing test/209/truth-2090p000.fits.
WARNING: UnitsWarning: The unit 'erg' has been deprecated in the FITS standard. Suggested: cm2 g s-2. [astropy.units.format.utils]
WARNING: UnitsWarning: The unit 'erg' has been deprecated in the FITS standard. Suggested: cm2 g s-2. [astropy.units.format.utils]
INFO:select_mock_targets:55:<module>: All done!

real    17m43.596s
user    371m28.817s
sys     18m5.516s
```

**Comment by moustakas** [comment]
OK, @sbailey, I think our packages have (re)converged to a reasonably stable state such that this PR is ~safe to merge . The only other updates needed at NERSC, after sourcing the `desi_environment.sh` script, are:
* We need a module for `desisim/0.18.0` (the latest tag) (and we need to test whether https://github.com/desihub/desisim/issues/216 is a persistent problem).
* `$DESI_BASIS_TEMPLATES` still points to the `v2.0` templates rather than `v2.3`

The code below ran successfully (in ~26m) over a 10 deg2 patch of sky on an interactive node on edison, indicating we're in a ~stable state.  Please poke around the output files and let me know what you think about the data model.  In the meantime, we should merge this and I'll continue working on the other high-priority items listed above, including scaling the code to run over the whole sky.
```
salloc -N 1 -p debug -L SCRATCH,project

source /project/projectdirs/desi/software/desi_environment.sh
module swap desiutil/1.9.3
module swap desispec/0.13.1

export TMPSCRATCH=/scratch1/scratchdirs/ioannis
cd $TMPSCRATCH
git clone git@github.com:desihub/desitarget.git
git clone git@github.com:desihub/desisim.git
cd desitarget ; git checkout spectra
cd ../desisim ; git checkout 0.18.0
for package in desitarget desisim; do                                                                         
  module unload $package                                                                                      
  export PATH=${TMPSCRATCH}/$package/bin:${PATH}                                                                 
  export PYTHONPATH=${TMPSCRATCH}/$package/py:${PYTHONPATH}                                                      
done                                                                                                          
export DESI_BASIS_TEMPLATES=${DESI_ROOT}/spectro/templates/basis_templates/v2.3

cd ${TMPSCRATCH}/datachallenge2017
time select_mock_targets -c mock_input.yaml -s 444 --nproc 24 --output_dir test
```

**Comment by sbailey** [comment]
FYI, I verified that this works with the as-installed desi code + this branch of desitarget at NERSC with:
```
source /project/projectdirs/desi/software/desi_environment.sh
module swap desiutil/1.9.3
module swap desispec/0.13.1
module swap desisim/0.18.1
module unload desitarget

mkdir -p $SCRATCH/desi/test && cd $SCRATCH/desi/test

git clone https://github.com/desihub/desitarget
cd desitarget && git checkout spectra && cd ..
export PYTHONPATH=`pwd`/desitarget/py:$PYTHONPATH
export PATH=`pwd`/desitarget/bin:$PATH

cp /project/projectdirs/desi/datachallenge/fullsurvey2017/mock_moustakas.yaml .
[edit to replace /Users/ioannis/research/projects/desi -> /project/projectdirs/desi]

mkdir targets
srun -N 1 -n 1 -c 24 -p debug -L SCRATCH,project \
time select_mock_targets -c mock_moustakas.yaml -s 444 --nproc 24 --output_dir targets
```

I'll merge now and we can proceed with other updates in separate PRs.

FYI, I initially tried:
```
srun -N 1 -p debug -L SCRATCH,project \
time select_mock_targets -c mock_moustakas.yaml -s 444 --nproc 24 --output_dir targets
```
and it timed out mid way through.  I'm not sure if the missing "-n 1 -c 24" enabled it to succeed the second time or whether faster reading of cached files the second time around made the difference.

**Pull request merged**